File: yntra-ios/YntraIOS/ViewModels/MessagingViewModel.swift
import SwiftUI
import yntra_core

@MainActor
class MessagingViewModel: ObservableObject {
    @Published var messages: [MessageItem] = []
    @Published var errorMessage: String? = nil
    
    private var observer: SwiftDbObserver?

    init() {
        loadMessages()
        
        self.observer = SwiftDbObserver { [weak self] in
            self?.loadMessages()
        }
        registerObserver(observer: self.observer!)
    }

    func loadMessages() {
        Task {
            do {
                self.messages = try await getMessages(requesterUserId: "user-1", userId: "user-1")
            } catch {
                self.errorMessage = error.localizedDescription
            }
        }
    }

    func markAsRead(messageId: String) {
        Task {
            do {
                try await markMessageRead(requesterUserId: "user-1", id: messageId)
            } catch {
                self.errorMessage = error.localizedDescription
            }
        }
    }

    func sendMessageToUser(receiverId: String, subject: String, body: String) {
        Task {
            do {
                _ = try await sendMessage(
                    requesterUserId: "user-1",
                    workspaceId: "workspace-1",
                    senderId: "user-1",
                    receiverId: receiverId,
                    teamId: nil,
                    subject: subject,
                    body: body
                )
            } catch {
                self.errorMessage = error.localizedDescription
            }
        }
    }

    func sendMessageToTeam(teamId: String, subject: String, body: String) {
        Task {
            do {
                _ = try await sendMessage(
                    requesterUserId: "user-1",
                    workspaceId: "workspace-1",
                    senderId: "user-1",
                    receiverId: nil,
                    teamId: teamId,
                    subject: subject,
                    body: body
                )
            } catch {
                self.errorMessage = error.localizedDescription
            }
        }
    }
}

File: yntra-android/app/src/main/java/com/yntra/app/viewmodels/MessagingViewModel.kt
package com.yntra.app.viewmodels

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch
import uniffi.yntra_core.*
import com.yntra.app.KotlinDbObserver

class MessagingViewModel : ViewModel() {
    private val _messages = MutableStateFlow<List<MessageItem>>(emptyList())
    val messages: StateFlow<List<MessageItem>> = _messages

    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage

    private val observer = KotlinDbObserver {
        loadMessages()
    }

    init {
        loadMessages()
        registerObserver(observer)
    }

    fun loadMessages() {
        viewModelScope.launch {
            try {
                _messages.value = getMessages(com.yntra.app.SessionManager.activeUserId, com.yntra.app.SessionManager.activeUserId)
            } catch (e: Exception) {
                _errorMessage.value = e.message
            }
        }
    }

    fun markAsRead(messageId: String) {
        viewModelScope.launch {
            try {
                markMessageRead(com.yntra.app.SessionManager.activeUserId, messageId)
            } catch (e: Exception) {
                _errorMessage.value = e.message
            }
        }
    }

    fun sendMessageToUser(receiverId: String, subject: String, body: String) {
        viewModelScope.launch {
            try {
                sendMessage(
                    requesterUserId = com.yntra.app.SessionManager.activeUserId,
                    workspaceId = com.yntra.app.SessionManager.activeWorkspaceId,
                    senderId = com.yntra.app.SessionManager.activeUserId,
                    receiverId = receiverId,
                    teamId = null,
                    subject = subject,
                    body = body
                )
            } catch (e: Exception) {
                _errorMessage.value = e.message
            }
        }
    }

    fun sendMessageToTeam(teamId: String, subject: String, body: String) {
        viewModelScope.launch {
            try {
                sendMessage(
                    requesterUserId = com.yntra.app.SessionManager.activeUserId,
                    workspaceId = com.yntra.app.SessionManager.activeWorkspaceId,
                    senderId = com.yntra.app.SessionManager.activeUserId,
                    receiverId = null,
                    teamId = teamId,
                    subject = subject,
                    body = body
                )
            } catch (e: Exception) {
                _errorMessage.value = e.message
            }
        }
    }

    override fun onCleared() {
        super.onCleared()
        clearObservers()
    }
}
